File: ts/src/pty.ts
import { CommandHandle } from './commands.js'
import { ConnectionConfig, type ConnectionOpts, type Username } from './connectionConfig.js'
import { ProcessFrame, ProcessSocket } from './processSocket.js'
import { DataPlaneClient, withQuery } from './transport.js'
import { SandboxError } from './errors.js'

export interface PtySize {
  cols: number
  rows: number
}

export interface PtyCreateOpts extends Pick<ConnectionOpts, 'requestTimeoutMs' | 'signal'> {
  cols?: number
  rows?: number
  size?: PtySize
  cmd?: string
  cwd?: string
  user?: Username
  envs?: Record<string, string>
  timeoutMs?: number
  onData?: (data: Uint8Array) => void | Promise<void>
  onPty?: (data: Uint8Array) => void | Promise<void>
}

export interface PtyConnectOpts {
  onData?: (data: Uint8Array) => void | Promise<void>
  timeoutMs?: number
  requestTimeoutMs?: number
}

/** PTY helper backed by the sandbox process WebSocket runtime. */
export class Pty {
  constructor(
    private readonly dataPlane: DataPlaneClient,
    private readonly config: ConnectionConfig
  ) {}

  /** Create an interactive shell PTY and return its live command handle. */
  async create(opts: PtyCreateOpts): Promise<CommandHandle> {
    const socket = await new ProcessSocket(
      this.dataPlane.baseUrl,
      this.dataPlane.token,
      '/runtime/v1/process',
      opts.requestTimeoutMs ?? this.config.requestTimeoutMs,
      this.config.headers
    ).connect()
    const envs = { TERM: 'xterm-256color', LANG: 'C.UTF-8', LC_ALL: 'C.UTF-8', ...(opts.envs ?? {}) }
    const size = opts.size ?? { cols: opts.cols ?? 80, rows: opts.rows ?? 24 }
    const args = opts.cmd === undefined ? ['-i', '-l'] : ['-l', '-c', opts.cmd]
    await socket.sendJson({
      type: 'start',
      cmd: '/bin/bash',
      args,
      cwd: opts.cwd,
      user: opts.user,
      environment: envs,
      envs,
      stdin: true,
      pty: { cols: size.cols, rows: size.rows },
      timeout_ms: opts.timeoutMs ?? 60_000,
    })
    const first = await nextStarted(socket)
    const pid = framePid(first)
    if (pid === undefined) throw new SandboxError('PTY started frame did not include pid')
    const reconnect = async (cursor: number) => this.openPtyStream(pid, cursor, opts)
    return new CommandHandle(pid, socket, () => this.kill(pid), withFirst(first, socket), undefined, undefined, opts.onData ?? opts.onPty, undefined, reconnect)
  }

  /** Connect to a running PTY by pid. */
  async connect(pid: number | string, opts: PtyConnectOpts = {}): Promise<CommandHandle> {
    const stream = await this.openPtyStream(pid, 0, opts)
    const reconnect = async (cursor: number) => this.openPtyStream(stream.actualPid, cursor, opts)
    return new CommandHandle(stream.actualPid, stream.socket, () => this.kill(stream.actualPid), stream.events, undefined, undefined, opts.onData, undefined, reconnect)
  }

  /** Send input bytes or text to a PTY. */
  async sendStdin(pid: number | string, data: string | Uint8Array, opts: PtyConnectOpts = {}): Promise<void> {
    const handle = await this.connect(pid, opts)
    try {
      await handle.sendStdin(data)
    } finally {
      await handle.disconnect()
    }
  }

  /** Send input bytes or text to a PTY. */
  async sendInput(pid: number | string, data: string | Uint8Array, opts: PtyConnectOpts = {}): Promise<void> {
    return this.sendStdin(pid, data, opts)
  }

  /** Resize a running PTY. */
  async resize(pid: number | string, size: PtySize, opts: PtyConnectOpts = {}): Promise<void> {
    const handle = await this.connect(pid, opts)
    try {
      await handle.resize(size)
    } finally {
      await handle.disconnect()
    }
  }

  /** Kill a running PTY. */
  async kill(pid: number | string, opts: Pick<ConnectionOpts, 'requestTimeoutMs' | 'signal'> = {}): Promise<boolean> {
    await this.dataPlane.deleteJson(withQuery(`/runtime/v1/process/${encodeURIComponent(String(pid))}`, {
      signal: 'SIGKILL',
      kill_group: true,
    }), {
      requestTimeoutMs: opts.requestTimeoutMs,
      signal: opts.signal,
    })
    return true
  }

  private async openPtyStream(pid: number | string, cursor: number, opts: PtyConnectOpts | PtyCreateOpts = {}): Promise<{
    actualPid: number | string
    socket: ProcessSocket
    events: AsyncIterable<ProcessFrame>
  }> {
    const socket = await new ProcessSocket(
      this.dataPlane.baseUrl,
      this.dataPlane.token,
      withQuery(`/runtime/v1/process/${encodeURIComponent(String(pid))}/connect`, { since: cursor }),
      opts.requestTimeoutMs ?? this.config.requestTimeoutMs,
      this.config.headers
    ).connect()
    const first = await nextStarted(socket)
    return {
      actualPid: framePid(first) ?? pid,
      socket,
      events: withFirst(first, socket),
    }
  }
}

async function nextStarted(events: AsyncIterable<ProcessFrame>): Promise<ProcessFrame> {
  for await (const frame of events) {
    if (frame.type === 'started') return frame
  }
  throw new SandboxError('PTY ended before started frame')
}

async function* withFirst(first: ProcessFrame, rest: AsyncIterable<ProcessFrame>) {
  yield first
  yield* rest
}

function framePid(frame: ProcessFrame): number | string | undefined {
  const process = frame.process && typeof frame.process === 'object' ? frame.process as Record<string, unknown> : {}
  const pid = frame.pid ?? process.pid ?? process.id
  return typeof pid === 'number' || typeof pid === 'string' ? pid : undefined
}

File: rust/src/pty.rs
use serde_json::Value;

use crate::commands::{CommandHandle, Commands};
use crate::error::{Error, Result};
use crate::process_socket::ProcessSocket;
use crate::transport::DataPlaneClient;

/// PTY terminal size.
#[derive(Clone, Copy, Debug)]
pub struct PtySize {
    /// Number of terminal columns.
    pub cols: u16,
    /// Number of terminal rows.
    pub rows: u16,
}

/// Options for creating a PTY.
#[derive(Clone, Debug)]
pub struct PtyCreateOptions {
    /// Terminal size.
    pub size: PtySize,
    /// Working directory.
    pub cwd: Option<String>,
    /// Environment variables.
    pub envs: serde_json::Map<String, Value>,
    /// Process timeout in milliseconds.
    pub timeout_ms: Option<u64>,
}

impl Default for PtyCreateOptions {
    fn default() -> Self {
        Self {
            size: PtySize { cols: 80, rows: 24 },
            cwd: None,
            envs: Default::default(),
            timeout_ms: Some(60_000),
        }
    }
}

/// PTY helper backed by the sandbox process WebSocket runtime.
#[derive(Clone)]
pub struct Pty {
    data_plane: DataPlaneClient,
}

impl Pty {
    pub(crate) fn new(data_plane: DataPlaneClient) -> Self {
        Self { data_plane }
    }

    /// Create an interactive shell PTY.
    pub async fn create(&self, opts: PtyCreateOptions) -> Result<CommandHandle> {
        let mut socket = ProcessSocket::connect(
            &self.data_plane.base_url,
            &self.data_plane.token,
            "/runtime/v1/process",
        )
        .await?;
        let mut envs = serde_json::Map::new();
        envs.insert("TERM".into(), Value::String("xterm-256color".into()));
        envs.insert("LANG".into(), Value::String("C.UTF-8".into()));
        envs.insert("LC_ALL".into(), Value::String("C.UTF-8".into()));
        envs.extend(opts.envs);
        socket
            .send_json(&serde_json::json!({
                "type": "start",
                "cmd": "/bin/bash",
                "args": ["-i", "-l"],
                "cwd": opts.cwd,
                "environment": envs.clone(),
                "envs": envs,
                "stdin": true,
                "pty": {"cols": opts.size.cols, "rows": opts.size.rows},
                "timeout_ms": opts.timeout_ms,
            }))
            .await?;
        let first = next_started(&mut socket).await?;
        let pid = frame_pid(&first)
            .ok_or_else(|| Error::Sandbox("PTY started frame did not include pid".into()))?;
        Ok(CommandHandle::new(
            pid,
            socket,
            Commands::new(self.data_plane.clone(), Default::default()),
        ))
    }

    /// Connect to a running PTY by process id.
    pub async fn connect(&self, pid: impl ToString) -> Result<CommandHandle> {
        let pid = pid.to_string();
        let mut socket = ProcessSocket::connect(
            &self.data_plane.base_url,
            &self.data_plane.token,
            &format!("/runtime/v1/process/{pid}/connect?since=0"),
        )
        .await?;
        let first = next_started(&mut socket).await?;
        let actual_pid = frame_pid(&first).unwrap_or(pid);
        Ok(CommandHandle::new(
            actual_pid,
            socket,
            Commands::new(self.data_plane.clone(), Default::default()),
        ))
    }

    /// Send input bytes to a PTY.
    pub async fn send_stdin(&self, pid: impl ToString, data: impl AsRef<[u8]>) -> Result<()> {
        let mut handle = self.connect(pid).await?;
        handle.send_stdin(data).await
    }

    /// Send input bytes to a PTY.
    pub async fn send_input(&self, pid: impl ToString, data: impl AsRef<[u8]>) -> Result<()> {
        self.send_stdin(pid, data).await
    }

    /// Resize a running PTY.
    pub async fn resize(&self, pid: impl ToString, size: PtySize) -> Result<()> {
        let mut handle = self.connect(pid).await?;
        handle.resize(size.cols, size.rows).await
    }

    /// Kill a running PTY.
    pub async fn kill(&self, pid: impl ToString) -> Result<bool> {
        self.data_plane
            .post_json(
                &format!("/runtime/v1/process/{}/signal", pid.to_string()),
                serde_json::json!({"signal": "SIGKILL"}),
            )
            .await?;
        Ok(true)
    }
}

async fn next_started(socket: &mut ProcessSocket) -> Result<Value> {
    while let Some(frame) = socket.next_frame().await? {
        if frame.get("type").and_then(Value::as_str) == Some("started") {
            return Ok(frame);
        }
    }
    Err(Error::Sandbox("PTY ended before started frame".into()))
}

fn frame_pid(frame: &Value) -> Option<String> {
    frame
        .get("pid")
        .or_else(|| frame.pointer("/process/pid"))
        .or_else(|| frame.pointer("/process/id"))
        .map(|value| {
            value
                .as_str()
                .map(ToOwned::to_owned)
                .unwrap_or_else(|| value.to_string())
        })
}
